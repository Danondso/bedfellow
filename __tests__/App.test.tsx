import React from 'react';
import { act, fireEvent, render, screen, waitFor } from '@testing-library/react-native';
import { authorize, AuthorizeResult } from 'react-native-app-auth';
import { Alert, Platform } from 'react-native';
import axios from 'axios';
import ImageColors, { AndroidImageColors, IOSImageColors } from 'react-native-image-colors';
import RootNavigation from '../src/screens';
import defaultPalette from '../src/theme/styles';
import ErrorBoundary from './helpers/components/ErrorBoundary';
import { happyPathApiHandler, whoSampledParsingApiHandler } from './helpers';

jest.mock('react-native-app-auth', () => ({
  authorize: jest.fn(),
}));

jest.spyOn(Alert, 'alert');

// if this is not here, we get open handles
jest.useFakeTimers();
jest.mock('axios');

const androidColors: AndroidImageColors = {
  platform: 'android',
  average: defaultPalette.primaryBackground,
  dominant: defaultPalette.primaryBackground100,
  darkMuted: defaultPalette.accent,
  darkVibrant: defaultPalette.error,
};

const iosColors: IOSImageColors = {
  platform: 'ios',
  background: defaultPalette.primaryBackground,
  primary: defaultPalette.primaryBackground100,
  secondary: defaultPalette.secondaryBackground,
  detail: defaultPalette.shadow,
  quality: undefined,
};

const mockedAxios = axios as jest.Mocked<typeof axios>;
const mockedImageColors = ImageColors as jest.Mocked<typeof ImageColors>;

describe('App Test Suite', () => {
  const authResult: AuthorizeResult = {
    accessToken: 'accessToken',
    accessTokenExpirationDate: '',
    idToken: 'id',
    refreshToken: 'refreshToken',
    tokenType: 'bearer',
    scopes: [],
    authorizationCode: 'codeywodey',
  };
  beforeEach(async () => {
    jest.resetAllMocks();
    // @ts-ignore
    authorize.mockResolvedValueOnce(authResult);
    mockedAxios.get.mockImplementation((url) => happyPathApiHandler(url));
    await render(
      <ErrorBoundary>
        <RootNavigation />
      </ErrorBoundary>
    );
  });

  it('renders login button text', () => {
    expect(screen.getByText('Login')).toBeTruthy();
  });

  describe('logs in, loads current track from spotify, retrieves samples from bedfellow-db-api', () => {
    describe('android tests', () => {
      it('queues samples', async () => {
        Platform.OS = 'android';

        mockedAxios.post.mockResolvedValueOnce({
          data: null,
          status: 204,
        });

        mockedImageColors.getColors.mockResolvedValue(androidColors);

        await act(async () => fireEvent.press(await screen.findByText('Login')));
        await act(async () => fireEvent.press(await screen.findByText('Ponderosa Twins Plus One')));

        await waitFor(() => {
          expect(screen.findByText('Ponderosa Twins Plus One')).toBeDefined();
          expect(screen.getByText('Queued Bound by Ponderosa Twins Plus One')).toBeDefined();
          expect(authorize).toHaveBeenCalledTimes(1);
          expect(mockedAxios.get).toHaveBeenCalledTimes(3);
          expect(mockedImageColors.getColors).toHaveBeenCalledTimes(1);
          expect(mockedAxios.post).toHaveBeenCalledTimes(1);
        });
      });

      it('fails to queue sample sourced from a TV show', async () => {
        Platform.OS = 'android';

        mockedImageColors.getColors.mockResolvedValue(androidColors);

        await act(async () => fireEvent.press(await screen.findByText('Login')));
        await act(async () => fireEvent.press(await screen.findByText('Martin (TV show)')));

        await waitFor(() => {
          expect(screen.findByText('Martin (TV show)')).toBeDefined();
          expect(screen.getByText('Cannot queue tv show')).toBeDefined();

          expect(authorize).toHaveBeenCalledTimes(1);
          expect(mockedAxios.get).toHaveBeenCalledTimes(2);
          expect(mockedImageColors.getColors).toHaveBeenCalledTimes(1);
          expect(mockedAxios.post).toHaveBeenCalledTimes(0);
        });
      });
    });

    describe('ios tests', () => {
      it('queues samples', async () => {
        Platform.OS = 'ios';

        mockedAxios.post.mockResolvedValueOnce({
          data: null,
          status: 204,
        });

        mockedImageColors.getColors.mockResolvedValue(iosColors);

        await act(async () => fireEvent.press(await screen.findByText('Login')));
        await act(async () => fireEvent.press(await screen.findByText('Ponderosa Twins Plus One')));

        await waitFor(() => {
          expect(screen.findByText('Ponderosa Twins Plus One')).toBeDefined();
          expect(screen.getByText('Queued Bound by Ponderosa Twins Plus One')).toBeDefined();
          expect(authorize).toHaveBeenCalledTimes(1);
          expect(mockedAxios.get).toHaveBeenCalledTimes(3);
          expect(mockedImageColors.getColors).toHaveBeenCalledTimes(1);
          expect(mockedAxios.post).toHaveBeenCalledTimes(1);
        });
      });
    });
  });

  it('logs in, loads current track from spotify, loads and parses whosampled page for track, and retrieves samples from bedfellow-db-api', async () => {
    Platform.OS = 'android';

    mockedAxios.get.mockImplementation((url) => whoSampledParsingApiHandler(url));

    // post to bedfellow-db-api
    mockedAxios.post.mockResolvedValueOnce({
      status: 204,
    });

    mockedImageColors.getColors.mockResolvedValue(androidColors);

    await act(async () => fireEvent.press(await screen.getByText('Login')));
    // TODO check redirectUri
    expect(authorize).toHaveBeenCalledTimes(1);
    expect(mockedAxios.get).toHaveBeenCalledTimes(9);
    expect(mockedImageColors.getColors).toHaveBeenCalledTimes(1);
    expect(screen.getByText('Ponderosa Twins Plus One')).toBeDefined();
  });

  it('invokes authorize function when pressing Login', async () => {
    Platform.OS = 'ios';

    mockedAxios.get.mockRejectedValueOnce({
      response: {
        status: 500,
      },
    });
    await act(async () => fireEvent.press(await screen.findByText('Login')));
    // TODO check redirectUri
    expect(authorize).toHaveBeenCalledTimes(1);
    expect(screen.getByText('Nothing playing currently.')).toBeDefined();
    expect(screen.getByText('No samples found.')).toBeDefined();
    expect(mockedImageColors.getColors).toHaveBeenCalledTimes(0);
    expect(mockedAxios.get).toHaveBeenCalledTimes(1);
  });

  describe('Login Failures', () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });
    it('shows login failure alert', async () => {
      // @ts-ignore
      authorize.mockRejectedValueOnce('Failed to login');
      await act(async () => fireEvent.press(await screen.findByText('Login')));
      expect(Alert.alert).toHaveBeenCalledWith('Failed to login');
    });
    it('shows login failure alert with error text', async () => {
      // @ts-ignore
      authorize.mockRejectedValueOnce(new Error('Oops'));
      await act(async () => fireEvent.press(await screen.findByText('Login')));
      expect(Alert.alert).toHaveBeenCalledWith('Failed to login', 'Oops');
    });
  });
});
