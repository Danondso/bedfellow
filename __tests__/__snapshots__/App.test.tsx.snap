// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders login screen snapshot 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RNSScreenStack
      style={
        {
          "flex": 1,
        }
      }
    >
      <RNSScreen
        collapsable={false}
        forwardedRef={[Function]}
        gestureResponseDistance={
          {
            "bottom": -1,
            "end": -1,
            "start": -1,
            "top": -1,
          }
        }
        nativeBackButtonDismissalEnabled={false}
        onAppear={[Function]}
        onDisappear={[Function]}
        onDismissed={[Function]}
        onHeaderBackButtonClicked={[Function]}
        onNativeDismissCancelled={[Function]}
        onTransitionProgress={[Function]}
        onWillDisappear={[Function]}
        replaceAnimation="push"
        stackPresentation="push"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        swipeDirection="horizontal"
      >
        <RNSScreenStackHeaderConfig
          backButtonInCustomView={false}
          backgroundColor="rgb(255, 255, 255)"
          color="rgb(0, 122, 255)"
          direction="ltr"
          disableBackButtonMenu={false}
          hidden={true}
          hideBackButton={false}
          largeTitleHideShadow={false}
          title="Login"
          titleColor="rgb(28, 28, 30)"
          topInsetEnabled={false}
          translucent={false}
        />
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          style={
            {
              "flex": 1,
              "flexDirection": "column-reverse",
            }
          }
        >
          <View
            style={
              [
                {
                  "flex": 1,
                },
                {
                  "backgroundColor": "rgb(242, 242, 242)",
                },
                undefined,
              ]
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#FBF2C4",
                  "flex": 1,
                  "justifyContent": "center",
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#E5C185",
                    "borderColor": undefined,
                    "borderRadius": 16,
                    "borderWidth": 1,
                    "paddingBottom": 200,
                    "width": "80%",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": undefined,
                      "fontSize": 48,
                      "paddingTop": 48,
                    }
                  }
                >
                   Bedfellow 
                </Text>
                <Text
                  style={
                    {
                      "color": undefined,
                      "fontSize": 16,
                      "textAlign": "center",
                    }
                  }
                >
                  a smol bean app
                </Text>
                <View
                  style={
                    {
                      "backgroundColor": "#FEF9E0",
                      "borderColor": undefined,
                      "borderRadius": 16,
                      "borderWidth": 1,
                      "marginTop": 32,
                    }
                  }
                >
                  <View
                    accessibilityState={
                      {
                        "busy": undefined,
                        "checked": undefined,
                        "disabled": undefined,
                        "expanded": undefined,
                        "selected": undefined,
                      }
                    }
                    accessibilityValue={
                      {
                        "max": undefined,
                        "min": undefined,
                        "now": undefined,
                        "text": undefined,
                      }
                    }
                    accessible={true}
                    collapsable={false}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      {
                        "borderColor": "#74A892",
                        "borderRadius": 16,
                        "opacity": 1,
                        "padding": 16,
                        "width": "100%",
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": undefined,
                          "fontSize": 32,
                          "fontWeight": "400",
                        }
                      }
                    >
                      Login
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RNSScreen>
    </RNSScreenStack>
  </RNCSafeAreaProvider>
</View>
`;
